docs: add blog post for js-ipfs 48 (#454)
ipfs/blog: content/post/104-js-ipfs-0.48.md

## 🧭 Delegate nodes on by default

JS IPFS has traditionally primarily targeted the browser, and the browser is a bad place to be if you want to be on the [DHT][]. You typically aren't on a page long enough to make or respond to [DHT][] queries and nor are you diallable so even if you were able to advertise yourself as a provider for a given block, the chances are no-one can connect to you to retrieve that block which results in a degraded service for everyone. Worse, the way you find more peers and content is via the [DHT][] so you're kind of stuck.

There are several ways to give in-browser IPFS nodes a better experience on the network, one of those is [Delegate Nodes](https://blog.ipfs.io/2019-08-06-js-ipfs-0-37/#delegated-peer-and-content-routing). A Delegate Node is a network peer that performs certain actions on behalf of other nodes on the network. In this case it will make [DHT][] queries on our behalf so we can find more peers and more content than ever before.

`js-IPFS@0.48.0` enables delegate nodes in the configuration by default. That means you should see far more peers that you have previously and be able to find content faster and more reliably than ever before.

> 

Reviewer: We should mention that the default config uses public delegate nodes, but in a production environment other delegate nodes should be setup to guarantee availability and avoid congestion. I have in my TODO list to create some documentation around libp2p infrastructure/deployment guidelines [libp2p/js-libp2p#696](https://github.com/libp2p/js-libp2p/issues/696) that will be useful for this. But meanwhile, I think we can just say that people should configure their own as we already have some information on how to do that, but it is spread across multiple channels

Author: Done!

There are several ways to give in-browser IPFS nodes a better experience on the network, one of those is [Delegate Nodes](https://blog.ipfs.io/2019-08-06-js-ipfs-0-37/#delegated-peer-and-content-routing). A Delegate Node is a network peer that performs certain actions on behalf of other nodes on the network. In this case it will make [DHT][] queries on our behalf so we can find more peers and more content than ever before.

`js-IPFS@0.48.0` enables delegate nodes in the configuration by default. That means you should see far more peers that you have previously and be able to find content faster and more reliably than ever before.

By default it uses [public delegate nodes](https://github.com/ipfs/js-ipfs/blob/master/packages/ipfs/docs/DELEGATE_ROUTERS.md) to give you the best out-of-the-box experience. These nodes are a shared commons but have no availability guarentees and are a potentially a source of resource contention. If you are deploying JS IPFS in a production environment you should stand up your own delegate nodes and [configure JS IPFS](https://github.com/ipfs/js-ipfs/blob/master/docs/CONFIG.md#delegates) accordingly.

> ## 🏓 DHT configuration

Author: AFAIK you can't run the node as a DHT server, it'll only ever be in client mode, but yes, will do - good to address expectations.

> ## 🏓 DHT configuration

Author: I've added another paragraph about this.

```

Just pass `--migrate` and your blockstore will be converted:

```console
$ jsipfs daemon --migrate
Initializing IPFS daemon...
js-ipfs version: 0.48.0
System version: x64/darwin
Node.js version: 12.16.1
Swarm listening on /ip4/127.0.0.1/tcp/4002/p2p/QmSRS11FZeMHvqe5wZamurpgj2Jmm9SYX76t7ZtJ7Tt74d
Swarm listening on /ip4/192.168.1.109/tcp/4002/p2p/QmSRS11FZeMHvqe5wZamurpgj2Jmm9SYX76t7ZtJ7Tt74d
Swarm listening on /ip4/127.0.0.1/tcp/4003/ws/p2p/QmSRS11FZeMHvqe5wZamurpgj2Jmm9SYX76t7ZtJ7Tt74d
API listening on /ip4/127.0.0.1/tcp/5002/http
Gateway (read only) listening on /ip4/127.0.0.1/tcp/9090/http
Web UI available at http://127.0.0.1:5002/webui
Daemon is ready
```

When used as a [module](https://github.com/ipfs/js-ipfs/blob/1760b8928dac14b3abcfa4a889042f0d7a956386/packages/ipfs/docs/MODULE.md) in the browser or an application [options.repoAutoMigrate](https://github.com/ipfs/js-ipfs/blob/1760b8928dac14b3abcfa4a889042f0d7a956386/packages/ipfs/docs/MODULE.md#optionsrepoautomigrate) is enabled by default so the upgrade will happen invisibly in the background; all you will notice is a one-off slightly longer startup time.

## 🗺️ A more intuitive API

> As the IPFS ecosystem grows more and more developers become interested in the project and start using our APIs. A lot of them have grown organically over time and not all of them have had equal amounts of thought put into them.

Reviewer 2: ```suggestion
As the IPFS ecosystem grows more and more developers become interested in the project and start using our APIs. A lot of them have grown organically over time and not all of them have had equal amounts of time invested in them.
```

We refactored the whole API from [callbacks](https://nodejs.org/en/knowledge/getting-started/control-flow/what-are-callbacks/) to [Promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise), then from returning [Arrays](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array) to [AsyncIterators](https://jakearchibald.com/2017/async-iterators-and-generators/) to allow streaming of enormous amounts of data without using external libraries.

In that last piece we lost a little bit of usability, as the humble `ipfs.add` API takes all manner of arguments and always returns an AsyncIterator:

```javascript
for await (const file of ipfs.add('My file content')) {
  // Wait, what?  I only added one file, why am I in a loop?
}
```

> With `js-IPFS@0.48.0`, `ipfs.add` now returns a single item. This seemingly innocuous change brings a raft of usability improvements as a very common question is 'I added a file, and got back `[object AsyncGenerator]`, what is that?' and then you have to get the whiteboard and the pens out and before you know it you've gone down an long explanation cul de sac when all they wanted to do was get a [CID][] back.

Reviewer 2: ```suggestion
With `js-IPFS@0.48.0`, `ipfs.add` now returns a single item. This seemingly innocuous change brings a raft of usability improvements as a very common question is 'I added a file, and got back `[object AsyncGenerator]`, what is that?' and then you have to get the whiteboard and the pens out and before you know it you've gone down a long explanation cul de sac when all they wanted to do was get a [CID][] back.
```

```javascript
const file = await ipfs.add('My file content')
// Ahh, much better
```

But you can still add whole directories of files if you want to, just call `ipfs.addAll`:

```javascript
const files = [{
  path: '/foo/file1.txt',
  content: 'file 1'
}, {
  path: '/foo/file2.txt',
  content: 'file 2'
}]

for await (const file of ipfs.addAll(files)) {
  // All the files!
}
```

### APIs with optional arguments

Recently we [released a change](https://blog.ipfs.io/2020-05-21-js-ipfs-0-44/#cancellable-requests) that allowed passing [AbortSignal][]s to all API methods. This necessitated adding an `options` object to every API call that didn't have one already. This left us in the weird situation where some arguments were optional, but were not in the options argument. Worse, the actions of some API calls changed dramatically depending on whether you passed an option or not. For example `ipfs.bootstrap.rm([multiaddr])` would completely empty the bootstrap list if you didn't pass a [Multiaddr][].

> All this leads to weird behaviour and subtle bugs when you pass things like `undefined` in for an optional arg position and don't pass an options argument, as well as knotty, error-prone internal code that tries to guess what you passed where based on type or properties of the objects where their types are the same.

Reviewer 2: ```suggestion
All this leads to weird behaviour and subtle bugs when you pass things like `undefined` in for an optional arg position and don't pass an options argument, as well as knotty, error-prone internal code that tries to guess what you passed based on type or properties of the objects where their types are the same.
```

With `js-IPFS@0.48.0`, all optional parameters to API methods now go in the options object. All APIs that have dramatic changes in behaviour have been split into more intuitive commands.

For example:

```javascript
// before
ipfs.bootstrap.add('/ip4/...') // adds a multiaddr to the bootstrap list
ipfs.bootstrap.add({ default: true }) // restores default bootstrappers

// after
ipfs.bootstrap.add('/ip4/...') // adds a multiaddr to the bootstrap list
ipfs.bootstrap.reset() // restores default bootstrappers
```

```javascript
// before
ipfs.bootstrap.rm('/ip4/...') // removes a multiaddr from the bootstrap list

> ipfs.bootstrap.rm({ all: true }) // restores default bootstrappers

Reviewer 2: ```suggestion
ipfs.bootstrap.rm({ all: true }) // empties bootstrapper list
```
